### PR TITLE
[SPARK-22895] [SQL] Push down the deterministic predicates that are after the first non-deterministic 

### PR DESCRIPTION
**Pull request by gatorsmile**
## What changes were proposed in this pull request?
Currently, we do not guarantee an order evaluation of conjuncts in either Filter or Join operator. This is also true to the mainstream RDBMS vendors like DB2 and MS SQL Server. Thus, we should also push down the deterministic predicates that are after the first non-deterministic, if possible.

## How was this patch tested?
Updated the existing test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -875,13 +875,15 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
...
-      // We could in theory span() only on determinism and pull out deterministic predicates
-      // on the watermark separately. But it seems unnecessary and a bit confusing to not simply
-      // use the prefix as we do for nondeterminism in other cases.
-
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).span(
-        p => p.deterministic && !p.references.contains(watermark.eventTime))
+      val (pushDown, stayUp) = {
+        val pushDownCondition: Expression => Boolean =
+          p => p.deterministic && !p.references.contains(watermark.eventTime)
+        if (SQLConf.get.outOfOrderPredicateEvaluationEnabled) {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Nit: Create a function which accepts parameter `Expression => Boolean` and `predicates: Seq[Expression]`, and `partitionByDeterminism` is using the function parameter as `e => e. deterministic`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala [hunk 2]
```diff
@@ -47,6 +48,15 @@ trait Predicate extends Expression {
...
 trait PredicateHelper {
+
+  protected def partitionByDeterminism(predicates: Seq[Expression]) = {
+    if (SQLConf.get.outOfOrderPredicateEvaluationEnabled) {
+      predicates.partition(_.deterministic)
+    } else {
+      predicates.span(_.deterministic)
+    }
+  }
+
```
Comment (attilapiros) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala]: Is not this "if" in the other way around here? As Seq#partition keeps the relative orders and Seq#span does not.
After reversing this "if" the "outOfOrderPredicateEvaluationEnabled" default value (true) will be consistent with the old behavior, before your PR (I mean using the Seq#span), but in the unit tests the relevant "if" conditions should be changed too.

**Comment by gatorsmile** [comment]
cc @cloud-fan @rxin @sameeragarwal @gengliangwang @jiangxb1987 @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 3]
```diff
@@ -768,6 +768,7 @@ object SQLConf {
     .booleanConf
     .createWithDefault(true)
 
+
```
Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: nit: duplicated empty line?

**Review by jiangxb1987 — APPROVED** [review]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -851,7 +851,7 @@ object PushDownPredicate extends Rule[LogicalPlan] with PredicateHelper {
 
     case filter @ Filter(condition, union: Union) =>
       // Union could change the rows, so non-deterministic predicate can't be pushed down
-      val (pushDown, stayUp) = splitConjunctivePredicates(condition).span(_.deterministic)
+      val (pushDown, stayUp) = splitConjunctivePredicates(condition).partition(_.deterministic)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: What does it mean "after the first non-deterministic"? Doesn't this simply partition predicates to deterministic and non-deterministic? Have it considered "first" non-deterministic?

Comment (jinxing64) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: @viirya 
IIUC, with `span`, the deterministic predicates after the first non-deterministic will not be pushed down, but with `partition`, all deterministic predicates will be pushed down.

**Pull request closed without merging**